### PR TITLE
schedule: Schedule plan interface and plan collector

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: ref #5257
should be merged after #5402 

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/schedule/plan/plan.go [hunk 1]
```diff
@@ -14,6 +14,78 @@
...
 // Plan is the basic unit for both scheduling and diagnosis.
 // TODO: for each scheduler/checker, we can have an individual definition but need to implement the common interfaces.
-type Plan interface{}
+type Plan interface {
+	GetSourceStore() *core.StoreInfo
+	GetRegion() *core.RegionInfo
+	GetTargetStore() *core.StoreInfo
+	SetSourceStore(store *core.StoreInfo)
+	SetRegion(region *core.RegionInfo)
+	SetTargetStore(store *core.StoreInfo)
```
Comment (rleungx) [inline server/schedule/plan/plan.go:29]: Do these interfaces fit all schedulers and checkers?

File: server/schedule/plan/plan.go [hunk 2]
```diff
@@ -14,6 +14,78 @@
...
+	GetStatus() Status
+	Step() int
+	Clone(...PlanOption) Plan
+}
+
+type PlanCollector struct {
+	basePlan           Plan
+	unschedulablePlans []Plan
+	schedulablePlans   []Plan
+	enable             bool
```
Comment (rleungx) [inline server/schedule/plan/plan.go:40]: prefer to remove it and use a dynamic config to control it.

File: server/schedulers/schedule_plan.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/schedule/plan"
+)
+
+type balanceSchedulerBasePlan struct {
+	source *core.StoreInfo
+	target *core.StoreInfo
+	region *core.RegionInfo
+	status plan.Status
+	step   int
```
Comment (rleungx) [inline server/schedulers/schedule_plan.go:30]: I don't think using int to describe the step is a good choice.

File: server/schedulers/balance_test.go [hunk 4]
```diff
@@ -67,7 +68,8 @@ func TestInfluenceAmp(t *testing.T) {
 	tc.AddLeaderRegion(1, 1, 2)
 	region := tc.GetRegion(1).Clone(core.SetApproximateSize(R))
 	tc.PutRegion(region)
-	plan := newBalancePlan(kind, tc, influence)
+	basePlan := NewBalanceSchedulerBasePlan()
+	plan := newSolver(basePlan, kind, tc, influence)
```
Comment (rleungx) [inline server/schedulers/balance_test.go:72]: solver?

File: server/schedule/plan/plan.go [hunk 5]
```diff
@@ -16,4 +16,69 @@ package plan
...
+
+	Clone(ops ...Option) Plan
+
+	// temp function
+	Desc() string
+}
+
+// Collector is a plan collector
+type Collector struct {
+	basePlan           Plan
```
[the remark this answers is not in the document]
Comment (CabinfeverB) [reply server/schedule/plan/plan.go:37]: To reduce the number of arguments passed in, I put the pointer to the plan to be cloned inside the struct. `base` refers to the source plan of the clone

File: server/schedulers/schedule_plan.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+		return "select region"
+	case 2:
+		return "select target store"
+	case 3:
+		return "verify the plan"
+	case 4:
+		return "create operator"
+	}
+	return ""
+}
```
Comment (rleungx) [inline server/schedulers/schedule_plan.go:38]: Should these be updated?

**Review by rleungx — CHANGES_REQUESTED** [review]

File: server/schedulers/schedule_plan.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+package schedulers
+
+import (
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/schedule/plan"
+)
+
+type balanceStep int
+
+func (b *balanceStep) Desc() string {
```
Comment (rleungx) [inline server/schedulers/schedule_plan.go:24]: Should this file need to be updated?

File: server/schedule/filter/filters.go [hunk 8]
```diff
@@ -49,14 +49,32 @@ func SelectSourceStores(stores []*core.StoreInfo, filters []Filter, opt *config.
 	})
 }
 
+// SelectSourceStoresWithCollector selects stores that be selected as source store from the list.
+func SelectSourceStoresWithCollector(stores []*core.StoreInfo, filters []Filter, opt *config.PersistOptions, collector *plan.Collector) []*core.StoreInfo {
```
Comment (rleungx) [inline server/schedule/filter/filters.go:53]: How about making it an optional parameter?

File: server/schedulers/schedule_plan.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	return int(*b)
+}
+
+func (b *balanceStep) Add() {
+	*b++
+}
+
+func (b *balanceStep) Sub() {
+	*b--
+}
```
Comment (rleungx) [inline server/schedulers/schedule_plan.go:34]: Is it necessary?

File: server/schedule/plan/plan.go [hunk 10]
```diff
@@ -16,4 +16,63 @@ package plan
 
 // Plan is the basic unit for both scheduling and diagnosis.
 // TODO: for each scheduler/checker, we can have an individual definition but need to implement the common interfaces.
-type Plan interface{}
+type Plan interface {
+	GetStep() int
+	GetStatus() *Status
+
+	Clone(ops ...Option) Plan // generate plan for clone option
+	GenerateCoreResource(interface{})
```
Comment (rleungx) [inline server/schedule/plan/plan.go:24]: What does core here mean?

File: server/schedule/filter/region_filters.go [hunk 11]
```diff
@@ -51,7 +69,8 @@ func SelectOneRegion(regions []*core.RegionInfo, filters ...RegionFilter) *core.
 // RegionFilter is an interface to filter region.
 type RegionFilter interface {
 	// Return true if the region can be used to schedule.
-	Select(region *core.RegionInfo) plan.Status
+	Select(region *core.RegionInfo) *plan.Status
+	Name() string
```
Comment (rleungx) [inline server/schedule/filter/region_filters.go:73]: Why do we need it?

File: server/schedule/filter/filters.go [hunk 12]
```diff
@@ -30,17 +30,26 @@ import (
 	"go.uber.org/zap"
 )
 
+type StoreFilterResultOption func(*plan.Status, *core.StoreInfo)
```
Comment (rleungx) [inline server/schedule/filter/filters.go:33]: How about making it shorter? BTW, why use the option?

File: server/schedule/plan/plan.go [hunk 13]
```diff
@@ -16,4 +16,63 @@ package plan
...
+// Collector is a plan collector
+type Collector struct {
+	basePlan           Plan
+	unschedulablePlans []Plan
+	schedulablePlans   []Plan
+	enable             bool
+}
+
+// NewCollector returns a new Collector
+func NewCollector(enable bool, plan Plan) *Collector {
```
Comment (rleungx) [inline server/schedule/plan/plan.go:37]: Do we need to create it if the dryrun is false?

Comment (rleungx) [inline server/schedule/plan/plan.go:37]: Do we need to create it if the dryrun is false?

File: server/schedule/filter/region_filters.go [hunk 14]
```diff
@@ -39,9 +47,19 @@ func filterRegionsBy(regions []*core.RegionInfo, keepPred func(*core.RegionInfo)
 }
 
 // SelectOneRegion selects one region that be selected from the list.
-func SelectOneRegion(regions []*core.RegionInfo, filters ...RegionFilter) *core.RegionInfo {
+func SelectOneRegion(regions []*core.RegionInfo, filters []RegionFilter, rops ...RegionFilterResultOption) *core.RegionInfo {
 	for _, r := range regions {
-		if slice.AllOf(filters, func(i int) bool { return filters[i].Select(r).IsOK() }) {
+		if filters == nil || slice.AllOf(filters,
```
Comment (rleungx) [inline server/schedule/filter/region_filters.go:52]: I suggest using `len` to check if there is a filter

File: server/schedule/filter/filters.go (hunk 8 above)
Comment (CabinfeverB) [reply server/schedule/filter/filters.go]: I want making collector be an optional parameter

File: server/schedule/filter/filters.go (hunk 12 above)
Comment (CabinfeverB) [reply server/schedule/filter/filters.go:33]: > How about making it shorter? BTW, why use the option?
to make collector be an optional parameter in `SelectXXX`

File: server/schedule/plan/plan.go (hunk 13 above)
Comment (CabinfeverB) [reply server/schedule/plan/plan.go:37]: I think we can create. The advantage is that we don't have to check nil in related code.

File: server/schedule/filter/region_filters.go [hunk 15]
```diff
@@ -51,7 +69,7 @@ func SelectOneRegion(regions []*core.RegionInfo, filters ...RegionFilter) *core.
 // RegionFilter is an interface to filter region.
 type RegionFilter interface {
 	// Return true if the region can be used to schedule.
```
Comment (rleungx) [inline server/schedule/filter/region_filters.go:71]: need to be updated

File: server/schedule/filter/region_filters.go [hunk 16]
```diff
@@ -20,8 +20,16 @@ import (
 	"github.com/tikv/pd/server/schedule/plan"
 )
 
+type RegionFilterResultOption func(*plan.Status, *core.RegionInfo)
+
+func CollectRegionPlan(collector *plan.Collector) func(*plan.Status, *core.RegionInfo) {
+	return func(status *plan.Status, r *core.RegionInfo) {
+		collector.Collect(plan.GenerateResource(r), plan.SetStatus(status))
+	}
+}
```
Comment (rleungx) [inline server/schedule/filter/region_filters.go:29]: Can we just pass a collecter if we don't have any other options?

File: server/schedule/plan/plan.go [hunk 17]
```diff
@@ -16,4 +16,63 @@ package plan
 
 // Plan is the basic unit for both scheduling and diagnosis.
 // TODO: for each scheduler/checker, we can have an individual definition but need to implement the common interfaces.
-type Plan interface{}
+type Plan interface {
+	GetStep() int
+	GetStatus() *Status
+
+	Clone(ops ...Option) Plan // generate plan for clone option
+	GenerateResource(interface{})
```
Comment (rleungx) [inline server/schedule/plan/plan.go:24]: How about using `SetXXX`?

File: server/schedule/plan/plan.go (hunk 13 above)
Comment (rleungx) [reply server/schedule/plan/plan.go:37]: Then, we need to create it every time `Schedule` is called.

File: server/schedulers/balance_region.go [hunk 18]
```diff
@@ -164,105 +167,118 @@ func (s *balanceRegionScheduler) Schedule(cluster schedule.Cluster, dryRun bool)
 	default:
 		baseRegionFilters = append(baseRegionFilters, filter.NewRegionEmptyFilter(cluster))
 	}
+	plusRegionFilters := append(baseRegionFilters, pendingFilter)
```
Comment (rleungx) [inline server/schedulers/balance_region.go:170]: plus?

File: server/schedulers/balance_region.go [hunk 19]
```diff
@@ -138,18 +138,29 @@ func (s *balanceRegionScheduler) IsScheduleAllowed(cluster schedule.Cluster) boo
 }
 
 func (s *balanceRegionScheduler) Schedule(cluster schedule.Cluster, dryRun bool) ([]*operator.Operator, []plan.Plan) {
+	basePlan := NewBalanceSchedulerPlan()
+	var collector *plan.Collector
+	var planFunc func() []plan.Plan
+	if dryRun {
+		collector = plan.NewCollector(basePlan)
+		planFunc = collector.GetPlans
```
Comment (rleungx) [inline server/schedulers/balance_region.go:146]: How about directly using `collector.GetPlans` as the return value?

File: server/schedulers/balance_region.go [hunk 20]
```diff
@@ -221,14 +215,20 @@ func (s *balanceRegionScheduler) Schedule(cluster schedule.Cluster, dryRun bool)
...
 			basePlan.step--
 		}
 		s.retryQuota.Attenuate(solver.source)
 	}
 	s.retryQuota.GC(stores)
-	return nil, planFunc()
+	if dryRun {
+		return nil, collector.GetPlans()
+	}
+	return nil, nil
```
Comment (rleungx) [inline server/schedulers/balance_region.go:231]: How about directly `return nil, collector.GetPlans()`

File: server/schedulers/balance_plan.go [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+	return basePlan
+}
+
+func (p *balanceSchedulerPlan) GetStep() int {
+	return p.step
+}
+
+func (p *balanceSchedulerPlan) SetResource(resource interface{}) {
+	switch p.step {
+	case 0:
```
Comment (rleungx) [inline server/schedulers/balance_plan.go:44]: Please add some comments to explain these numbers

File: server/schedulers/balance_region.go [hunk 22]
```diff
@@ -165,104 +170,128 @@ func (s *balanceRegionScheduler) Schedule(cluster schedule.Cluster, dryRun bool)
 		baseRegionFilters = append(baseRegionFilters, filter.NewRegionEmptyFilter(cluster))
 	}
 
-	for _, plan.source = range stores {
-		retryLimit := s.retryQuota.GetLimit(plan.source)
+	basePlan.step++
```
Comment (rleungx) [inline server/schedulers/balance_region.go:173]: Can we unify the usage for solver and plan? in `transferPeer`, it is solver.step++/--

**Review by rleungx — COMMENTED** [review]
The rest LGTM

**Review by rleungx — APPROVED** [review]

File: server/cluster/cluster_test.go [hunk 23]
```diff
@@ -1685,10 +1685,10 @@ func Test(t *testing.T) {
 	pendingFilter := filter.NewRegionPendingFilter()
 	downFilter := filter.NewRegionDownFilter()
 	for i := uint64(0); i < n; i++ {
-		region := filter.SelectOneRegion(tc.RandLeaderRegions(i, []core.KeyRange{core.NewKeyRange("", "")}), pendingFilter, downFilter)
+		region := filter.SelectOneRegion(tc.RandLeaderRegions(i, []core.KeyRange{core.NewKeyRange("", "")}), nil, pendingFilter, downFilter)
```
Comment (nolouch) [inline server/cluster/cluster_test.go:1688]: It's better to replace the `nil` with a `mock collector` in all tests.

**Review by nolouch — COMMENTED** [review]
I think the `step`  may have better way to calculate. maybe improve it in next PR.
rest LGTM

File: server/schedulers/balance_plan.go [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+
+// NewBalanceSchedulerPlan returns a new balanceSchedulerBasePlan
+func NewBalanceSchedulerPlan() *balanceSchedulerPlan {
+	basePlan := &balanceSchedulerPlan{
+		status: plan.NewStatus(plan.StatusOK),
+	}
+	return basePlan
+}
+
+func (p *balanceSchedulerPlan) GetStep() int {
```
[the remark this answers is not in the document]
Comment (CabinfeverB) [reply server/schedulers/balance_plan.go:38]: plan is built in `Schedule` function, and there is no concurrency problem because `Schedule` is serial.

**Review by nolouch — APPROVED** [review]

**Comment by nolouch** [comment]
@CabinfeverB can merge it?

**Pull request merged**